### PR TITLE
(Not ready for publication) Web Fundamentals Update post: Unified Media Pipeline

### PR DESCRIPTION
**Pull request by samdutton**
Can squash the commits on merge :).

Google docs version was reviewed by @jpmedley and media folks.

### CODE REVIEW

**Comment by TripleCxx** [comment]
https://github.com/googlebot
https://github.com/google/WebFundamentals/labels/cla%3A%20yes

**Comment by TripleCxx** [comment]
12

File: src/content/en/updates/posts/2016/04/ump.markdown [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+There are a couple of bugs affecting &lt;video&gt;, service workers and the Cache Storage API:
+
+* [&lt;video&gt; triggers a mode: cors request and won't accept an opaque service worker response](https://bugs.chromium.org/p/chromium/issues/detail?id=546076)
+* [Seeking doesn't work in videos served by service worker cache](https://bugs.chromium.org/p/chromium/issues/detail?id=575357)
+
+
+## Browser support
+
+* Behind a flag in Chrome Dev since January 2016
```
Comment (jeffposnick) [inline src/content/en/updates/posts/2016/04/ump.markdown]: M51 will be stable at around May 31, so you can leave out the bit about the flag and Chrome Dev.

File: src/content/en/updates/posts/2016/04/ump.markdown [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+* Pass MediaStreams between Web Audio and MediaRecorder.
+* Develop and maintain media apps more easily across devices — media works the same on desktop and Android.
+
+UMP took some hard engineering work to implement:
+
+* A new caching layer for improved power performance.
+* Updating a new MediaCodec based video decoder hosted in Chrome's GPU process.
+* Lots of testing and iteration on different devices.
+
+Here's a <a href="https://googlechrome.github.io/samples/service-worker/prefetch-video/index.html">demo of video caching with a service worker</a>:
```
Comment (jeffposnick) [inline src/content/en/updates/posts/2016/04/ump.markdown]: See https://github.com/GoogleChrome/samples/pull/336#issuecomment-221276991

**Comment by jeffposnick** [comment]
This should be fine from my perspective once https://github.com/GoogleChrome/samples/pull/336#issuecomment-221276991 is resolved in the related sample code.

File: src/content/en/updates/posts/2016/04/ump.markdown (hunk 1 above)
Comment (samdutton) [inline src/content/en/updates/posts/2016/04/ump.markdown]: UMP has just been pushed back, so I think enabled by default from 52 is correct now.

File: src/content/en/updates/posts/2016/04/ump.markdown [hunk 3]
```diff
@@ -0,0 +1,156 @@
+---
+layout: updates/post
+title: "Service worker caching, playbackRate and Blob URLS for audio and video on Chrome for Android"
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/04/ump.markdown]: Lower-case B on blob.

File: src/content/en/updates/posts/2016/04/ump.markdown (hunk 2 above)
Comment (samdutton) [inline src/content/en/updates/posts/2016/04/ump.markdown]: Updated sample is now live (but with questions about video hosting).

File: src/content/en/updates/posts/2016/04/ump.markdown (hunk 3 above)
Comment (samdutton) [inline src/content/en/updates/posts/2016/04/ump.markdown]: Done.

File: src/content/en/updates/posts/2016/04/ump.markdown [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+Case in point: the Unified Media Pipeline, **UMP** for short.
+
+This may sound like a sinister Soviet era directive, but in fact it's an important step towards consistent cross-platform audio and video delivery. Chrome on Android will now use the same media stack as desktop Chrome, rather than relying on the underlying platform implementation.
+
+UMP enables you to do a lot:
+
+* Cache audio and video with service workers, since media delivery is now implemented directly within Chrome rather than being passed off to the Android media stack.
+* Use Blob URLs for audio and video elements.
+* Set `playbackRate` for audio and video.
+* Pass MediaStreams between Web Audio and MediaRecorder.
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/04/ump.markdown]: Don't capitalize Web Audio.

File: src/content/en/updates/posts/2016/04/ump.markdown [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+      reader.onloadend = function() {
+        video.src = URL.createObjectURL(new Blob([this.result]));
+      };
+      reader.readAsArrayBuffer(file);
+    }, handleError);
+  }, handleError);
+}
+{% endhighlight %}
+
+The Unified Media Pipeline has also been [enabled for MSE and EME](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-reviews/Qi4dLcKjcCM).
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/04/ump.markdown]: Spell out MSE and EME.

File: src/content/en/updates/posts/2016/04/ump.markdown [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+This is another step towards unifying mobile and desktop Chrome. You don't need to change your code, but building a consistent media experience across desktop and mobile should now be easier, since the media stack is the same across platforms. Debugging with Chrome DevTools? Mobile emulation now uses the 'real' audio and video stack.
+
+If you experience problems as a result of the Unified Media Pipeline, please file issues on the [implementation bug](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-reviews/Qi4dLcKjcCM) or via [new.crbug.com](https://new.crbug.com).
+
+## Demos
+
+* <a href="https://googlechrome.github.io/samples/service-worker/prefetch-video/index.html">Cache video with a service worker</a>
+* <a href="https://simpl.info/video/playbackrate">Media `playbackRate`</a>
+* <a href="https://simpl.info/mediarecorder">MediaRecorder: playback using a Blob URL</a>
+* <a href="https://simpl.info/video/offline">Offline video implemented with the File APIs</a>
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/04/ump.markdown]: Don't capitalize file.

File: src/content/en/updates/posts/2016/04/ump.markdown (hunk 4 above)
Comment (samdutton) [inline src/content/en/updates/posts/2016/04/ump.markdown]: Web Audio (capitalized) refers to the Web Audio API (not just 'web audio').

**Comment by jpmedley** [comment]
I made a few editorial remarks.

File: src/content/en/updates/posts/2016/04/ump.markdown (hunk 6 above)
Comment (samdutton) [inline src/content/en/updates/posts/2016/04/ump.markdown]: This is the name of a set of APIs: the File APIs. (See [simpl.info/mse](https://simpl.info/mse/) for an example.)

**Comment by samdutton** [comment]
Thanks Joe!

**Pull request merged**